run-make-support: tidy up support library

jieyouxu opened the pull request:
- Make `handle_failed_output` take `&Command` instead of having the caller keep doing `format!("{:#?}", s)`.
- Introduce a helper macro for implementing common command wrappers, such as `arg`, `args`, `run`, `run_fail`.
    - Use the helper macro on existing command wrappers and remove manual copy-pasta'd implementations.

jieyouxu committed:
run-make-support: make `handle_failed_output` take a `&Command`

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8622647654/job/23634223291) [(plain)](https://github.com/rust-lang/rust/commit/7ab8d733452b1bff32757f2f2525aee8dcebf210/checks/23634223291/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 29.2s done
#17 DONE 35.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.40s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
```

</details>

jieyouxu commented:
@rustbot author

jieyouxu commented:
@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #123708) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fbstj commented on src/tools/run-make-support/src/lib.rs:
```diff
@@ -115,3 +115,99 @@ fn handle_failed_output(cmd: &Command, output: Output, caller_line_number: u32)
     std::process::exit(1)
 }
+
+macro_rules! impl_common_helpers {
```
is it worth documenting (or just commenting) what is delegated for developers at least?

jieyouxu replied:
I'll make sure to add some docs

jieyouxu replied:
Thanks for the reminder, I added some docs to explain what this macro provides.

jieyouxu committed:
run-make-support: use macro to implement common methods

Removes the manual copy-pasta'd implementation of common methods.

jieyouxu committed:
run-make-support: add some top-level docs

bors commented:
:umbrella: The latest upstream changes (presumably #123838) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Mark-Simulacrum commented:
r=me with a rebase

jieyouxu committed:
run-make-support: introduce macro for common methods to avoid repetition

Add a helper macro for adding common methods to command wrappers. Common
methods include helpers that delegate to `Command` and running methods.

- `arg` and `args` (delegates to `Command`)
- `env`, `env_remove` and `env_clear` (delegates to `Command`)
- `output`, `run` and `run_fail`

This helps to avoid needing to copy-pasta / reimplement these common
methods on a new command wrapper, which hopefully reduces the friction
for run-make test writers wanting to introduce new command wrappers.

jieyouxu commented:
Rebased and fixed the merge conflict.
@bors r=Mark-Simulacrum rollup

bors commented:
:pushpin: Commit a67a119424377ec53a3b36bca251a30637ae42dd has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum a67a119424377ec53a3b36bca251a30637ae42dd -->
<!-- homu: {"type":"Approved","sha":"a67a119424377ec53a3b36bca251a30637ae42dd","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
